### PR TITLE
[SPARK-20918] [SQL] Use FunctionIdentifier as function identifiers in FunctionRegistry

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
Currently, the unquoted string of a function identifier is being used as the function identifier in the function registry. This could cause the incorrect the behavior when users use `.` in the function names. This PR is to take the `FunctionIdentifier` as the identifier in the function registry. 

- Add one new function `createOrReplaceTempFunction` to `FunctionRegistry`
```Scala
final def createOrReplaceTempFunction(name: String, builder: FunctionBuilder): Unit
```

### How was this patch tested?
Add extra test cases to verify the inclusive bug fixes.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 1]
```diff
@@ -72,39 +89,53 @@ trait FunctionRegistry {
...
-      functionBuilders.get(name).map(_._2).getOrElse {
+      functionBuilders.get(normalizeFuncName(name)).map(_._2).getOrElse {
         throw new AnalysisException(s"undefined function $name")
       }
     }
     func(children)
   }
 
-  override def listFunction(): Seq[String] = synchronized {
-    functionBuilders.iterator.map(_._1).toList.sorted
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: This `sorted` is useless.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 2]
```diff
@@ -72,39 +89,53 @@ trait FunctionRegistry {
 
 class SimpleFunctionRegistry extends FunctionRegistry {
 
-  protected val functionBuilders =
-    StringKeyHashMap[(ExpressionInfo, FunctionBuilder)](caseSensitive = false)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: This has a bug. The database name could be case sensitive.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 3]
```diff
@@ -1116,8 +1115,8 @@ class SessionCatalog(
     // TODO: just make function registry take in FunctionIdentifier instead of duplicating this
     val database = name.database.orElse(Some(currentDb)).map(formatDatabaseName)
     val qualifiedName = name.copy(database = database)
-    functionRegistry.lookupFunction(name.funcName)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: This also sounds a bug. This line ignores the database name.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 4]
```diff
@@ -1205,8 +1204,8 @@ class SessionCatalog(
     requireDbExists(dbName)
     val dbFunctions = externalCatalog.listFunctions(dbName, pattern).map { f =>
       FunctionIdentifier(f, Some(dbName)) }
-    val loadedFunctions =
-      StringUtils.filterPattern(functionRegistry.listFunction(), pattern).map { f =>
+    val loadedFunctions = StringUtils
+      .filterPattern(functionRegistry.listFunction().map(_.unquotedString), pattern).map { f =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: This PR keeps the current behavior. However, I think it is also a bug. The user-specified `pattern` should not consider the database name.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 5]
```diff
@@ -17,51 +17,74 @@
...
   }
 
-  def registerFunction(name: String, info: ExpressionInfo, builder: FunctionBuilder): Unit
+  def registerFunction(
+    name: FunctionIdentifier,
+    info: ExpressionInfo,
+    builder: FunctionBuilder): Unit
+
+  /* Create or replace a temporary function. */
+  final def createOrReplaceTempFunction(name: String, builder: FunctionBuilder): Unit = {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Since we already expose `FunctionRegistry` to the stable class `UDFRegistration`, I added this extra API for a helper function. 

Ideally, this function should only exist in `SessionCatalog`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 6]
```diff
@@ -1152,27 +1151,27 @@ class SessionCatalog(
...
-      return functionRegistry.lookupFunction(qualifiedName.unquotedString, children)
+      return functionRegistry.lookupFunction(qualifiedName, children)
     }
 
     // The function has not been loaded to the function registry, which means
     // that the function is a permanent function (if it actually has been registered
     // in the metastore). We need to first put the function in the FunctionRegistry.
     // TODO: why not just check whether the function exists first?
     val catalogFunction = try {
-      externalCatalog.getFunction(currentDb, name.funcName)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: This is another bug that blocks users to use the function qualified for the other database. For example, 
```SQL
SELECT db1.test_avg(1)
```

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Submitted a separate PR #18146 for easily backporting to the previous branch.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 7]
```diff
@@ -72,39 +95,53 @@ trait FunctionRegistry {
 
 class SimpleFunctionRegistry extends FunctionRegistry {
 
-  protected val functionBuilders =
-    StringKeyHashMap[(ExpressionInfo, FunctionBuilder)](caseSensitive = false)
+  @GuardedBy("this")
+  private val functionBuilders =
+    new mutable.HashMap[FunctionIdentifier, (ExpressionInfo, FunctionBuilder)]
+
+  // Resolution of the function name is always case insensitive, but the database name
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: this looks weird, database name is always case sensitive and function name is always case insenstive?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: I think sorted output can make users easy to search for a function, shall we still keep it?

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 8]
```diff
@@ -61,7 +61,7 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
         | dataType: ${udf.dataType}
       """.stripMargin)
 
-    functionRegistry.registerFunction(name, udf.builder)
+    functionRegistry.createOrReplaceTempFunction(name, udf.builder)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: is it same as before? what if the `name` contains database part?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: That is the resolution rule we are using now. : (

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala (hunk 8 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: Our current register function APIs in UDFRegistration does not allow users to specify the database names as part of `name`, since the registered functions are temporary.

**Comment by gatorsmile** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: we can fix it as a follow-up

**Comment by cloud-fan** [comment]
LGTM, merging to master!

**Comment by rxin** [comment]
Guys - please in the future separate bug fixes with refactoring. Don't mix a bunch of cosmetic changes with actual bug fixes together.

**Comment by HyukjinKwon** [comment]
I am leaving a note (at least to myself) since it looked anyhow caused behaviour change.


With this statements in Hive side:

```sql
CREATE TABLE emp AS SELECT 'user' AS name, 'address' as address;
CREATE DATABASE d100;
CREATE FUNCTION d100.udf100 AS 'org.apache.hadoop.hive.ql.udf.generic.GenericUDFUpper';
```


**Hive**


```
hive> SELECT d100.udf100(`emp`.`name`) FROM `emp`;
USER
hive> SELECT `d100.udf100`(`emp`.`name`) FROM `emp`;
USER
```


**Spark**

Before:

```
scala> spark.sql("SELECT d100.udf100(`emp`.`name`) FROM `emp`").show
+-----------------+
|d100.udf100(name)|
+-----------------+
|             USER|
+-----------------+


scala> spark.sql("SELECT `d100.udf100`(`emp`.`name`) FROM `emp`").show
+-----------------+
|d100.udf100(name)|
+-----------------+
|             USER|
+-----------------+
```

After:

```
scala> spark.sql("SELECT d100.udf100(`emp`.`name`) FROM `emp`").show
+-----------------+
|d100.udf100(name)|
+-----------------+
|             USER|
+-----------------+


scala> spark.sql("SELECT `d100.udf100`(`emp`.`name`) FROM `emp`").show
org.apache.spark.sql.AnalysisException: Undefined function: 'd100.udf100'. This function is neither a registered temporary function nor a permanent function registered in the database 'default'.; line 1 pos 7
```


**MySQL**

This change causes a inconsistency with Hive although looks consistent compating to MySQL.

```
mysql> SELECT `d100.udf100`(`emp`.`name`) FROM `emp`;
ERROR 1305 (42000): FUNCTION hkwon.d100.udf100 does not exist
mysql> SELECT d100.udf100(`emp`.`name`) FROM `emp`;
+---------------------------+
| d100.udf100(`emp`.`name`) |
+---------------------------+
| Hello, user!              |
+---------------------------+
1 row in set (0.01 sec)
```

**Comment by cloud-fan** [comment]
@HyukjinKwon Thanks for the note! I think this behavior is better, I'm adding a `release_note` tag to the JIRA ticket, so that we don't forget to mention it in release notes.

**Comment by HyukjinKwon** [comment]
I agree with this change too for clarification.

**Comment by dongjoon-hyun** [comment]
BTW, this is changed at Spark 2.3.0. How did we handle this before?

**Comment by cloud-fan** [comment]
hmm, then it's too late. Maybe we can add it in Spark 2.3.2, cc @jerryshao

**Comment by jerryshao** [comment]
I see. Thanks for the note.

**Comment by HyukjinKwon** [comment]
@cloud-fan, should we update migration guide as well?

**Comment by cloud-fan** [comment]
After a second thought, isn't it a bug?
```
hive> SELECT `d100.udf100`(`emp`.`name`) FROM `emp`;
USER
```

This clearly violates the SQL semantic: the string inside backticks should be treated as a string literal. I think we should update the JIRA ticket to explain this bug, but we don't need to put it in a release notes or migration guide.

I'll update the ticket.

**Comment by HyukjinKwon** [comment]
> This clearly violates the SQL semantic: the string inside backticks should be treated as a string literal.

BTW, I believe there's no particular standard for backticks themselves since different DBMS uses different backtick implementations.

**Comment by HyukjinKwon** [comment]
One explicit problem here is, we claim Hive compatibility in Spark. The difference should be explained when we are clear on this.

**Comment by cloud-fan** [comment]
> BTW, I believe there's no particular standard for backticks themselves since different DBMS uses different backtick implementations.

You are right, but SQL standard does define how to quote identifiers, by using `"xyz"`. Databases usually support one more syntax to quote identifiers, e.g. backticks, square brackets, etc.

So for this case, I think it's obvious that users want to quote the function name, and we have a bug.

> One explicit problem here is, we claim Hive compatibility in Spark.

Do we? Sometimes we follow hive behaviors, but we never guarantee that IIRC.

**Comment by gatorsmile** [comment]
We do not need to follow Hive if Hive does not follow SQL compliance. Our main goal is to follow the mainstream DBMS vendors. 

BTW, we can enhance our parser to recognize the other symbols (e.g., double quotes) as the quotes instead of forcing users to choose backtick. cc @maropu

**Comment by HyukjinKwon** [comment]
I mean https://spark.apache.org/docs/latest/sql-programming-guide.html#supported-hive-features and https://spark.apache.org/docs/latest/sql-programming-guide.html#unsupported-hive-functionality

The issue is also related with VIEW support as well. Should be good to note.

I don't mean we should necessarily follow Hive's behaviour for clarification .. Also, I agree with this change ..

**Comment by cloud-fan** [comment]
> Spark SQL is designed to be compatible with the Hive Metastore, SerDes and UDFs.

This is different from `Spark can run any Hive SQL`. Spark can load and use Hive UDFs, with the right SQL syntax.

Anyway I'm happy to see a PR to note it. It's good to be verbose in the doc. But I won't treat it a must-have.

**Comment by HyukjinKwon** [comment]
Yea, I didn't mean it super seriously @cloud-fan - I just left a comment in case for a better documentation since I see many users go from Hive to Spark.

**Comment by maropu** [comment]
@gatorsmile btw, we've already file a lira to track these kinds of all the SQL-compiliant issues? not yet?
